util.c

```diff
@@ -39,38 +48,88 @@ void safe_buffer_free(struct safe_buffer * sb)
 	}
 }
 
-size_t safe_buffer_pad(struct safe_buffer * sb, size_t count)
-{
-	size_t rc = 0;
-	unsigned char zero = 0;
+void safe_buffer_check_used_size(size_t proposed_size) {
+	if (proposed_size > 64*1024) {
+		flog(LOG_ERR, "Requested buffer too large for any possible IPv6 ND, even with jumbogram.  Exiting.");
+		exit(1);
+	}
+}
 
-	while (count--) {
-		rc += safe_buffer_append(sb, &zero, 1);
+void safe_buffer_resize(struct safe_buffer * sb, size_t new_capacity)
+{
+	// This may slightly exceed 64KiB at some points.
+	if (sb->allocated < new_capacity) {
+		sb->allocated = new_capacity;
+		sb->buffer = realloc(sb->buffer, sb->allocated);
 	}
+}
+
+void safe_buffer_expand(struct safe_buffer * sb, size_t additional_capacity)
+{
+	safe_buffer_resize(sb, sb->allocated + additional_capacity);
+}
 
-	return rc;
+size_t safe_buffer_pad(struct safe_buffer * sb, size_t count)
+{
+	safe_buffer_check_used_size(sb->used + count);
+	safe_buffer_expand(sb, count);
+	memset(&sb->buffer[sb->used], (uint8_t)0, count);
+	sb->used += count;
+	return count;
 }
 
 size_t safe_buffer_append(struct safe_buffer * sb, void const * v, size_t count)
 {
 	if (sb) {
+		safe_buffer_check_used_size(sb->used + count);
 		unsigned const char * m = (unsigned const char *)v;
-		if (sb->allocated <= sb->used + count) {
-			sb->allocated = sb->used + count + MSG_SIZE_SEND;
-			sb->buffer = realloc(sb->buffer, sb->allocated);
-		}
+		safe_buffer_expand(sb, count);
 		memcpy(&sb->buffer[sb->used], m, count);
 		sb->used += count;
-
-		if (sb->used >= MSG_SIZE_SEND) {
-			flog(LOG_ERR, "Too many prefixes, routes, rdnss or dnssl to fit in buffer.  Exiting.");
-			exit(1);
-		}
 	}
 
 	return count;
 }
 
+struct safe_buffer_list * new_safe_buffer_list(void)
+{
+	struct safe_buffer_list * sbl = malloc(sizeof(struct safe_buffer_list));
+	sbl->sb = new_safe_buffer();
+	sbl->next = NULL;
+	return sbl;
+}
+
+struct safe_buffer_list * safe_buffer_list_append(struct safe_buffer_list * sbl)
+{
+	// Only allocate a new entry if this one has bytes in it.
+	if(sbl->sb && sbl->sb->used > 0) {
+		struct safe_buffer_list * next = new_safe_buffer_list();
+		sbl->next = next;
```

Comment:
I'd consider throwing in `assert(sbl->next == NULL)` at the start, because it really should be for appending to the list. The reason all the new functions return `struct safe_buffer_list *`, is that they return the new tail of the list.

Using `ADD_TO_LL` would mostly be not returning the tail of the list anymore, and instead always passing in the head. I wondered about this as a functional change, but it would mean a lot of seeking to the end of the list.